python/file_generator.py

```diff
@@ -0,0 +1,377 @@
+		"""Creates a clean file or a set of edit files specified by the
+		function call"""
+
+		#Creates TS row data. 
+		self.ts_row = self.make_ts_row()
+
+		#Creates a TS row dataframe. 
+		ts_df = pd.DataFrame(self.ts_row, index=[1])
+
+		#The following produces a clean file. 
+		if kind == 'clean_file':
+
+			#Creates a first row and a second row of LAR data which result in 
+			#a LAR dataframe for which more rows are added. 
+			for i in range(0, self.file_length):
+				print('Creating row {i}'.format(i=i))
+				if i==0:
+					lar_row = self.make_clean_lar_row(ts_row=self.ts_row)
+					first_row = pd.DataFrame(lar_row, index=[1])
+				elif i==1:
+					new_row = self.make_clean_lar_row(ts_row=self.ts_row)
+					new_lar = pd.DataFrame(new_row, index=[1])
+					lar_frame = pd.concat([first_row, new_lar], axis=0)
+				else:
+					new_row = self.make_clean_lar_row(ts_row=self.ts_row)
+					new_lar = pd.DataFrame(new_row, index=[1])
+					lar_frame = pd.concat([lar_frame, new_lar], axis=0)
+			
+			#Writes the file to a clean filepath specified in the test_filepaths
+			#configuration.  
+			utils.write_file(ts_input=ts_df, lar_input=lar_frame, 
+				path=self.filepaths['clean_filepath'].format(
+					bank_name=self.data_map["name"]["value"]),
+				name=self.filepaths['clean_filename'].format(
+					n=self.file_length, bank_name=self.data_map["name"]["value"]))
+
+		#For error files. 
+		if kind == 'error_files':
+
+			#Modifies clean data and outputs 
+			#resulting files that fail specific edits.
+			
+			#Instantiates the edit file maker.
+			file_maker = test_data(ts_schema=self.ts_schema_df, 
+				lar_schema=self.lar_schema_df) 
+
+			#Pulls in the clean data filepath and name from the
+			#test filepaths yaml file. 
+			ts_data, lar_data = utils.read_data_file(
+				path=self.filepaths['clean_filepath'].format(
+					bank_name=self.data_map["name"]["value"]),
+				data_file=self.filepaths["clean_filename"].format(
+					bank_name=self.data_map["name"]["value"], 
+					n=self.data_map["file_length"]["value"])) 
+					
+
+			#Passes clean file data to the file maker object.
+			file_maker.load_data_frames(ts_data, lar_data) 
+
+			#Generates a file for each edit function in file maker. 
+			edits = []
+			
+			#Loops over all data modification functions. 
+			for func in dir(file_maker): 
+
+				#Checks if function is a numbered syntax or validity edit.
+				if func[:1] in ("s", "v", "q") and func[1:4].isdigit()==True: 
+					print("applying:", func)
+					#Applies data modification functions and produces files.
+					getattr(file_maker, func)() 
+					
+
+	def edit_report(self, data_filepath, data_filename):
+		"""
+		This function takes in a filepath and name, producing a report on
+		whether any rows of the data failed syntax, validity or quality edits.
+
+		The report contains among its fields the edit name, the status of the 
+		edit, the number of rows failed by the edit, if any, and the ULI's of 
+		NULIs of the rows that fail the edit. 
```

Comment:
Can you add (loan ID) after NULIs? Just so there's a clear translation.

Reply:
Sure! This change has been made.